File: 2-3/01KnapsackProblem2.py
#coding: UTF-8

MAX_N = 100
MAX_W = 10000

n =4
W = 5
w = [2, 1, 3, 2]
v = [3, 2, 4, 2]

# メモ化テーブル
dp = [[-1 for j in range(MAX_W+1)] for i in range(MAX_N)]

# i番目以降の品物から重さの総和がj以下になるように選ぶ
def rec(i, j):
    # 調べたことがある場合はテーブルの値を返す
    if dp[i][j] >= 0:
        return dp[i][j]
    
    if i == n:
        # all item is evaluated
        res = 0
    elif(j < w[i]):
        # i番目の品物は入らない
        res = rec(i + 1, j)
    else:
        # 入れる場合、入れない場合のうち価値が高い方を選択
        res = max(rec(i + 1, j), rec(i + 1, j - w[i]) + v[i])
    dp[i][j] = res
    return res

def solve():
    print "the answer is " + str(rec(0, W))


solve()

File: 2-3/01KnapsackProblem2.cpp
#include<iostream>

const int MAX_N = 100;
const int MAX_W = 10000;

int n = 4;
int W = 5;
int w[MAX_N] = {2, 1, 3, 2,};
int v[MAX_N] = {3, 2, 4, 2,};

int dp[MAX_N+1][MAX_W+1];

int max(int a, int b){
	return ((a>b)? a:b);
}

int rec(int i, int j){
	if(dp[i][j] >= 0){
		return dp[i][j];
	}
	
	int res;
	if(i==n){
		res = 0;
	}else if(j < w[i]){
		res = rec(i + i, j);
	}else{
		res = max(rec(i + 1, j), rec(i + 1, j - w[i]) + v[i]);
	}
	return dp[i][j] = res;
}

void solve(){
	memset(dp, -1, sizeof(dp));
	std::cout << rec(0, W) << std::endl;
}

int main(){
	solve();
	return 0;
}
